# Linear Regression

### 1. Import thư viện cần thiết và load dataset

In [65]:
import os
import joblib
import pandas as pd
from sklearn.linear_model import LinearRegression

data_train = pd.read_csv("../data/preprocessed/data_train.csv")
data_test = pd.read_csv("../data/preprocessed/data_test.csv")

display(data_train)
display(data_test)

,distance_m,signal_mean_pct,rx_link_rate_mean,rtt_mean_ms,jitter_ms,packet_loss_pct,throughput_down_mbps
0,5,90.000000,144.4,1.500000,0.777778,0.0,88.205492
1,15,33.250000,1.0,108.888889,55.500000,10.0,0.000000
2,15,34.000000,1.0,290.500000,262.222222,0.0,0.105660
3,1,84.230769,866.7,1.000000,0.000000,0.0,93.401740
4,1,83.000000,866.7,2.111111,2.500000,10.0,94.606291
...,...,...,...,...,...,...,...
283,1,83.000000,866.7,2.500000,2.333333,0.0,92.199988
284,5,89.333333,144.4,3.222222,4.625000,10.0,84.029303
285,5,90.000000,144.4,1.300000,0.444444,0.0,79.932116
286,1,84.615385,866.7,1.000000,0.000000,0.0,92.998248


,distance_m,signal_mean_pct,rx_link_rate_mean,rtt_mean_ms,jitter_ms,packet_loss_pct,throughput_down_mbps
0,5,91.000000,144.400000,1.5,1.111111,0.0,90.725926
1,15,41.000000,5.250000,3.1,3.222222,0.0,0.000000
2,5,87.750000,144.400000,1.4,0.777778,0.0,76.034358
3,15,29.000000,1.000000,992.6,882.555556,0.0,0.000000
4,5,84.833333,144.400000,5.8,6.777778,0.0,18.118921
...,...,...,...,...,...,...,...
67,5,89.000000,144.400000,2.4,2.000000,0.0,89.400531
68,15,30.000000,1.000000,158.2,148.444444,0.0,0.000000
69,15,24.250000,1.000000,159.2,211.777778,0.0,0.000000
70,15,49.500000,2.708333,361.4,333.888889,0.0,0.418837


### 2. Định nghĩa các biến `FEATURES` và biến `TARGET`

In [66]:
FEATURES = [
    "distance_m",
    "signal_mean_pct",
    "rx_link_rate_mean",
    "rtt_mean_ms",
    "jitter_ms",
    "packet_loss_pct"
]

TARGET = "throughput_down_mbps"

X_train = data_train[FEATURES]
y_train = data_train[TARGET]

X_test = data_test[FEATURES]
y_test = data_test[TARGET]

### 3. Train mô hình

In [67]:
model = LinearRegression()
model.fit(X_train, y_train)

y_intercept = model.intercept_
coeff_table = pd.DataFrame({
    'Feature': model.feature_names_in_,
    'Coefficient': model.coef_
})

print("The y-intercept of the model is:", y_intercept)
display(coeff_table)

The y-intercept of the model is: -31.504708394830992


,Feature,Coefficient
0,distance_m,-0.021684
1,signal_mean_pct,1.093367
2,rx_link_rate_mean,0.040383
3,rtt_mean_ms,-0.005455
4,jitter_ms,-0.001275
5,packet_loss_pct,-0.105101


#### Ý nghĩa
- `distance_m`: Cứ mỗi 1 mét đi xa khỏi Access Point, thông lượng sẽ mất đi ~0.022 Mbps
- `signal_mean_pct`: Cường độ tín hiệu tăng thêm 1% thì thông lượng sẽ tăng thêm ~1.093 Mbps
- `rx_link_rate`: Tốc độ liên kết tăng thêm 1 Mbps thì thông lượng sẽ tăng thêm ~0.04 Mbps
- `rtt_mean_ms`: Thời gian trễ trọn vòng tăng thêm 1 mili-giây thì thông lượng sẽ giảm đi ~0.005 Mbps
- `jitter_ms`: Biến thiên độ trễ tăng thêm 1 mili-giây thì thông lượng sẽ giảm đi ~0.001 Mbps
- `packet_loss_pct`: Tỷ lệ mất gói tin tăng thêm 1% thì thông lượng sẽ giảm đi ~0.105 Mbps

### 4. Sử dụng mô hình đã train để dự đoán biến `TARGET` của tập `data_test`

In [68]:
y_pred_linear_regression = model.predict(X_test)

y_pred_linear_regression = pd.DataFrame(
    y_pred_linear_regression,
    columns=["prediction_label"],
    index=X_test.index
)

display(y_pred_linear_regression)

,prediction_label
0,73.704923
1,13.189056
2,70.152452
3,-6.621818
4,66.931815
...,...
67,71.512147
68,-0.040826
69,-6.413875
70,20.003917


### 5. Lưu lại dự đoán của mô hình Linear Regression

In [69]:
try:
    y_pred_linear_regression.to_csv("../data/prediction-linear-regression/prediction_linear_regression.csv", mode="x", index=False)
    print("File created successfully.")
except FileExistsError:
    print("File already exists.")

File created successfully.


### 6. Lưu lại mô hình

In [70]:
model_path = "../models/linear_regression_model.joblib"

if not (os.path.exists(model_path)):
    joblib.dump(model, model_path)
    print("Model created successfully.")
else:
    print("File already exists.")

File already exists.
